### PR TITLE
feat: implement doctor reports feature

### PR DESCRIPTION
**Pull request by ryanMMM**
Introduced the reports feature, which allows doctors to write a report about a specific patient after loading the patient. Added the feature commands to the doctor loaded patient controller, added report presenter views, and fixed some of the bugs in report manager.

### CODE REVIEW

**Review by ahmedkhalf — CHANGES_REQUESTED** [review]
Please pull from the latest main branch and resolve the merge conflict on your computer.

**Review by DanielDervishi — DISMISSED** [review]
Looks Good To Me!

File: src/controllers/DoctorLoadedPatientController.java [hunk 1]
```diff
@@ -96,27 +96,39 @@ private Command DeletePatientPrescription() {
...
-        return (x) -> new ReportManager(getDatabase()).getReportData(patientData);
+    private Command ViewPatientReports() {
+        return (x) -> {
+            ArrayList<ReportData> reportData = reportManager.getReportData(patientData);
+            if (reportData.size() > 0) {
+                doctorView.viewAllReports(reportData);
+                Integer targetReport = doctorView.viewReportPrompt();
+                doctorView.viewReport(reportData.get(targetReport - 1));
+            } else {
+                doctorView.noReports();
```
Comment (ahmedkhalf) [inline src/controllers/DoctorLoadedPatientController.java]: Please follow the naming convention of presenters:
`showNoReportsError`

File: src/controllers/DoctorLoadedPatientController.java [hunk 2]
```diff
@@ -96,27 +96,39 @@ private Command DeletePatientPrescription() {
...
-    private Command deleteReport() {
+    private Command DeletePatientReport() {
         return (x) -> {
+            ArrayList<ReportData> reportData = reportManager.getReportData(patientData);
+
             Integer deleteIndex = doctorView.deleteReportPrompt(new ContactManager(getDatabase())
-                    .getContactData(patientData), new ReportManager(getDatabase()).getReportData(patientData));
-            getDatabase().getReportDatabase().remove(deleteIndex);
+                    .getContactData(patientData), reportData);
+            reportManager.deleteReport(reportData.get(deleteIndex));
```
Comment (ahmedkhalf) [inline src/controllers/DoctorLoadedPatientController.java]: What happens when the index is out of range? The user should be made aware of his mistake. What happens when the user input is not an integer? The user should be made aware of his mistake. Look into the secretary deletion code for guidance. Furthermore, have you manually tested your code? Without even testing, this code clearly looks like it's prone to null pointer errors.

File: src/presenters/screenViews/DoctorScreenView.java [hunk 3]
```diff
@@ -266,4 +280,21 @@ public void showDeleteAbsenceNotAnIntegerError() {
...
+    public void viewAllReports(ArrayList<ReportData> reportData) {
+        infoMessage(reportView.viewAsEnumerationFromList(reportData, reportView::viewHeader));
+    }
+
+    public Integer viewReportPrompt() {
+        return inputInt("Enter the index of the report to view in detail ");
+    }
+
+    public void noReports() {
+        infoMessage("No reports to view, please add a report");
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/DoctorScreenView.java]: This should be an error message.

File: src/controllers/DoctorLoadedPatientController.java [hunk 4]
```diff
@@ -96,27 +96,39 @@ private Command DeletePatientPrescription() {
...
-                doctorView.reportDetailsPrompt().header(), doctorView.reportDetailsPrompt().body());
+    private Command CreatePatientReport() {
+        return (x) -> {
+            ReportDetails reportDetails = doctorView.reportDetailsPrompt();
+            reportManager.addReport(patientData, doctorData, reportDetails.header(), reportDetails.body());
+        };
     }
 
-    private Command deleteReport() {
+    private Command DeletePatientReport() {
```
Comment (ahmedkhalf) [inline src/controllers/DoctorLoadedPatientController.java:140]: The user is not getting any feedback as to whether the deletion was successful or not.

File: src/presenters/screenViews/DoctorScreenView.java [hunk 5]
```diff
@@ -266,4 +280,21 @@ public void showDeleteAbsenceNotAnIntegerError() {
...
+
+    public void viewAllReports(ArrayList<ReportData> reportData) {
+        infoMessage(reportView.viewAsEnumerationFromList(reportData, reportView::viewHeader));
+    }
+
+    public Integer viewReportPrompt() {
+        return inputInt("Enter the index of the report to view in detail ");
+    }
+
+    public void noReports() {
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/DoctorScreenView.java]:
```suggestion
    public void showNoReportsError() {
```

File: src/useCases/ReportManager.java [hunk 6]
```diff
@@ -22,43 +20,45 @@ public class ReportManager {
...
      * Delete a report from the report database.
+     *
      * @param reportData ReportData - data associated with the report to be deleted.
-     * @return boolean - true if the id associated with the reportData exists in the database. false otherwise.
      */
-    public boolean deleteReport(ReportData reportData){
-        return reportDatabase.remove(reportData.getReportId());
+    public void deleteReport(ReportData reportData) {
+        reportDatabase.remove(reportData.getReportId());
     }
```
Comment (ahmedkhalf) [inline src/useCases/ReportManager.java:40]: Why did you remove the boolean return here? Could you explain this design decision?

File: src/useCases/ReportManager.java [hunk 7]
```diff
@@ -22,43 +20,45 @@ public class ReportManager {
...
     /**
      * Add a report to the report database. Assumes all patientData and doctorData represents a valid patient and
      * doctor within the database.
+     *
      * @param patientData PatientData - data associated with a patient.
      * @param doctorData  DoctorData - data associated with a doctor.
-     * @param header String - header of the report.
-     * @param body String - body of the report.
+     * @param header      String - header of the report.
+     * @param body        String - body of the report.
```
Comment (ahmedkhalf) [inline src/useCases/ReportManager.java]: Extraneous spaces.
```suggestion
     * @param header String - header of the report.
     * @param body String - body of the report.
```

**Review by ahmedkhalf — CHANGES_REQUESTED** [review]
@Hepaul7 Please code with good conscience in mind.

> Looks Good To Me!

@DanielDervishi What exactly looks good to you?

**Comment by DanielDervishi** [comment]
> @Hepaul7 Please code with good conscience in mind.
> 
> > Looks Good To Me!
> 
> @DanielDervishi What exactly looks good to you?

More review was required clearly. Will get on that.

File: src/presenters/screenViews/DoctorScreenView.java (hunk 3 above)
Comment (DanielDervishi) [reply src/presenters/screenViews/DoctorScreenView.java]: I don't believe this qualifies as an error. We are requesting to view all reports and there are none there. This is a not an un-desired behaviour.

File: src/controllers/DoctorLoadedPatientController.java [hunk 8]
```diff
@@ -54,13 +57,10 @@ public LinkedHashMap<String, Command> AllCommands() {
         LinkedHashMap<String, Command> commands = new LinkedHashMap<>();
         commands.put("unload patient", Back(previousController));
         commands.put("back", Back(previousController));
```
Comment (DanielDervishi) [inline src/controllers/DoctorLoadedPatientController.java:59]: It might be better to put commands related to the doctors actions first, and then the commands for changing state after.

**Review by DanielDervishi — CHANGES_REQUESTED** [review]
It might be better to put commands related to the doctors actions first, and then the commands for changing state after.

**Review by DanielDervishi — CHANGES_REQUESTED** [review]
When viewing a report in its entirety, you should display the name of the doctor who created this report. This is useful when a doctor views the reports of a patient that was previously not theirs.

File: src/controllers/DoctorLoadedPatientController.java (hunk 2 above)
Comment (Hepaul7) [reply src/controllers/DoctorLoadedPatientController.java]: I have manually tested my code and so far it works. I will however fix the presenter conventions and possible mistakes users could make.

File: src/useCases/ReportManager.java (hunk 6 above)
Comment (Hepaul7) [reply src/useCases/ReportManager.java:40]: I did not remove that, I believe Ryan made that change after.

**Review by ahmedkhalf — APPROVED** [review]

File: src/presenters/entityViews/ReportView.java [hunk 9]
```diff
@@ -44,4 +45,13 @@ public String viewBody(ReportData item) {
...
     }
 
+    /**
+
+     * @param item ReportData bundle to view.
+     * @return String representation of DoctorId as a view.
+     */
+    public String viewDoctor(ReportData item) {
+        return "Report was noted by doctor ID: " + item.getDoctorId();
+    }
```
Comment (DanielDervishi) [inline src/presenters/entityViews/ReportView.java]: This should not be displayed to the user.

File: src/controllers/DoctorLoadedPatientController.java [hunk 10]
```diff
@@ -86,37 +87,78 @@ private Command DeletePatientPrescription() {
...
+                return;
+            }
+            reportManager.deleteReport(reportData.get(deleteIndex));
+            doctorView.showReportDeletionSuccess();
+
         };
-    } */
+    }
 
+    private boolean CheckUserInteger(Integer deleteIndex, Integer size) {
```
Comment (DanielDervishi) [inline src/controllers/DoctorLoadedPatientController.java]: Please rename this method to something more descriptive like "isInvalidIndex" or something along those lines.

File: src/presenters/entityViews/ReportView.java (hunk 9 above)
Comment (DanielDervishi) [reply src/presenters/entityViews/ReportView.java]: Rather the name of the doctor who created the report should be displayed.

File: src/useCases/ReportManager.java [hunk 11]
```diff
@@ -62,4 +64,12 @@ public ArrayList<ReportData> getReportData(PatientData patientData) {
...
     }
 
+    public DoctorData getReportDoctor(ReportData report) {
+        Doctor doctor = doctorDatabase.get(report.getDoctorId());
+        if (doctor == null) {
+            return null;
+        }
+        return new DoctorData(doctor);
+    }
+
```
Comment (DanielDervishi) [inline src/useCases/ReportManager.java]: This method should probably exist in DoctorManager, where it takes in a doctor id rather than a reportData. When called in the controller, simply extract the id from the report data.

Comment (ahmedkhalf) [reply src/useCases/ReportManager.java]: You are correct!

**Review by DanielDervishi — APPROVED** [review]
Looks good to me.
